hedera-node/src/main/java/com/hedera/services/store/TypedTokenStore.java

```diff
@@ -102,19 +105,64 @@ public TypedTokenStore(
 			Supplier<FCMap<MerkleUniqueTokenId, MerkleUniqueToken>> uniqueTokens,
 			Supplier<FCMap<MerkleEntityAssociation, MerkleTokenRelStatus>> tokenRels,
 			BackingTokenRels backingTokenRels,
-			UniqTokenViewsManager uniqTokenViewsManager,
-			LegacyTreasuryAdder legacyStoreDelegate,
-			LegacyTreasuryRemover delegate
+			UniqTokenViewsManager uniqTokenViewsManager
 	) {
 		this.tokens = tokens;
 		this.uniqTokenViewsManager = uniqTokenViewsManager;
 		this.tokenRels = tokenRels;
 		this.uniqueTokens = uniqueTokens;
 		this.accountStore = accountStore;
 		this.transactionRecordService = transactionRecordService;
-		this.delegate = delegate;
 		this.backingTokenRels = backingTokenRels;
-		this.addKnownTreasury = legacyStoreDelegate;
+	}
+
+	public void rebuildViews() {
+		knownTreasuries.clear();
+		rebuildViewOfKnownTreasuries();
+	}
+
+	private void rebuildViewOfKnownTreasuries() {
+		forEach(tokens.get(), (key, value) -> {
+			/* A deleted token's treasury is no longer bound by ACCOUNT_IS_TREASURY restrictions. */
+			if (!value.isDeleted()) {
+				addKnownTreasury(Id.fromGrpcAccount(value.treasury().toGrpcAccountId()), Id.fromGrpcToken(key.toTokenId()));
+			}
+		});
+	}
+
+	public void addKnownTreasury(final Id aId, final Id tId) {
+		knownTreasuries.computeIfAbsent(aId, ignore -> new HashSet<>()).add(tId);
+	}
+
+	public void removeKnownTreasuryForToken(final Id aId, final Id tId) {
+		throwIfKnownTreasuryIsMissing(aId);
+		knownTreasuries.get(aId).remove(tId);
+		if (knownTreasuries.get(aId).isEmpty()) {
+			knownTreasuries.remove(aId);
+		}
+	}
+
+	private void throwIfKnownTreasuryIsMissing(final Id aId) {
+		if (!knownTreasuries.containsKey(aId)) {
+			throw new IllegalArgumentException(String.format(
+					"Argument 'aId=%s' does not refer to a known treasury!",
+					readableId(aId)));
+		}
+	}
+
+	public boolean isKnownTreasury(final Id aid) {
+		return knownTreasuries.containsKey(aid);
+	}
+
+	public boolean isTreasuryForToken(final Id aId, final Id tId) {
+		if (!knownTreasuries.containsKey(aId)) {
+			return false;
+		}
+		return knownTreasuries.get(aId).contains(tId);
+	}
+
+	Map<Id, Set<Id>> getKnownTreasuries() {
+		return knownTreasuries;
 	}
```

Comment:
nit: add java on public methods doc

Reply:
In the perspective of the approach wherein this PR, we will not include removing the knownTreasuries collection from the HederaTokenStore - we will follow this comment during the transfer of this collection in future discrete PR.